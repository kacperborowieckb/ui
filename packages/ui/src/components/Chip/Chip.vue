<template>
  <div
    :tabindex="isFocusable ? 0 : undefined"
    :class="chipVariants({ variant })"
    @keydown.backspace="handleChipRemoval"
  >
    <span class="flex gap-2 py-[3px] leading-none">
      <template v-if="content">
        {{ content }}
      </template>
      <slot />
    </span>
    <button
      v-if="withButton"
      tabindex="-1"
      class="cursor-pointer"
      @click="handleChipRemoval"
    >
      <Cross2Icon />
    </button>
  </div>
</template>

<script lang="ts">
import { Cross2Icon } from '@radix-icons/vue'
import { tv } from 'tailwind-variants'
import { computed } from 'vue'
import type { VariantProps } from 'tailwind-variants'

const chipVariants = tv({
  base: 'focus-visible:ring-offset-background focus-visible:ring-ring flex items-center gap-1 rounded-sm border px-2 transition-colors focus-visible:outline-none focus-visible:ring-2 ',
  variants: {
    variant: {
      default: 'bg-muted dark:bg-muted/80 dark:hover:bg-input text-secondary-foreground hover:bg-input/75 border-input',
      error: 'bg-destructive/10 border-destructive/20 text-destructive hover:bg-destructive/20 dark:brightness-150',
    },
  },
})

type ChipVariants = VariantProps<typeof chipVariants>

export interface ChipProps {
  content?: string
  variant?: ChipVariants['variant']
  focusable?: boolean
  disabled?: boolean
  withButton?: boolean
}

export interface ChipEmits {
  (e: 'chipRemoval'): void
}
</script>

<script setup lang="ts">
const props = withDefaults(defineProps<ChipProps>(), {
  variant: 'default',
})

const emit = defineEmits<ChipEmits>()

const isFocusable = computed(() => props.focusable && !props.disabled)

function handleChipRemoval() {
  emit('chipRemoval')
}
</script>
