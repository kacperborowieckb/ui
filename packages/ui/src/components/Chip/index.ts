export {
  default as Chip,
  type ChipEmits,
  type ChipProps,
} from './Chip.vue'
