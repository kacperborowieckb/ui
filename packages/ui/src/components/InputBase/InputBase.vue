<template>
  <input
    v-model="model"
    class="placeholder:text-muted-foreground py-1/2 h-full min-w-[120px] flex-1 px-2 focus:outline-0"
    type="text"
    :placeholder="placeholder"
    :disabled="disabled"
  >
</template>

<script lang="ts">
export interface InputBaseProps {
  placeholder?: string
  disabled?: boolean
}
</script>

<script setup lang="ts">
defineProps<InputBaseProps>()

const model = defineModel<string>()
</script>
