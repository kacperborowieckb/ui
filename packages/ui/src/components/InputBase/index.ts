export {
  default as InputBase,
  type InputBaseProps,
} from './InputBase.vue'
