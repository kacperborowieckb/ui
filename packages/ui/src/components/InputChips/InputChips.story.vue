<template>
  <Story>
    <Variant title="Default">
      <div class="max-w-[424px] p-4">
        <InputChips
          v-model="chips"
          label="Usernames to send email"
          placeholder="Username.."
          errorMessage="Usernames must contains 4-16 characters"
          :chipSeparators="['Tab', 'Enter', ',']"
          :validator="(val) => val.length >= 4 && val.length <= 16"
        />
      </div>
    </Variant>
    <Variant title="Error State">
      <div class="max-w-[424px] p-4">
        <InputChips
          hasError
          label="Usernames to send email"
          placeholder="Username.."
          :chipSeparators="['Tab', 'Enter', ',']"
        />
      </div>
    </Variant>
    <Variant title="Disabled">
      <div class="max-w-[424px] p-4">
        <InputChips
          disabled
          label="Usernames to send email"
          placeholder="Username.."
          :chipSeparators="['Tab', 'Enter', ',']"
        />
      </div>
    </Variant>
  </Story>
</template>

<script setup lang="ts">
import { ref } from 'vue'

import InputChips from './InputChips.vue'

const chips = ref<string[]>([])
</script>
