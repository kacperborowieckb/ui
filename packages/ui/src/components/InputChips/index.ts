export { default as InputChips } from './InputChips.vue'
