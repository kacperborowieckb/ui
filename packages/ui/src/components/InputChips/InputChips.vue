<template>
  <InputWrapper
    :label="label"
    :disabled="disabled"
    :hasError="hasErrors"
    :errorMessage="errorMessage"
  >
    <Chip
      v-for="(chip, index) in chips"
      :key="`${index}-${chip}`"
      focusable
      withButton
      :disabled="disabled"
      :content="chip"
      :variant="chipsWithErrors.has(chip) ? 'error' : 'default'"
      @chipRemoval="handleChipRemoval(chip, index)"
    />
    <InputBase
      v-model.trim="inputValue"
      :placeholder="placeholder"
      :disabled="disabled"
      @keydown="handleKeyDown"
    />
  </InputWrapper>
</template>

<script lang="ts">
import { computed, onMounted, ref, watch } from 'vue'

import { Chip } from '../Chip'
import { InputBase } from '../InputBase'
import { InputWrapper } from '../InputWrapper'

export interface InputChipsProps {
  placeholder?: string
  label?: string
  disabled?: boolean
  chipSeparators: string[]
  validator?: RegExp | ((chipValue: string) => boolean)
  hasError?: boolean
  errorMessage?: string
}

export interface InputChipsEmits {
  (e: 'chipsUpdated', chips: string[]): void
  (e: 'errorStateChanged', hasError: boolean): void
}
</script>

<script setup lang="ts">
const props = defineProps<InputChipsProps>()
const emit = defineEmits<InputChipsEmits>()

const chips = defineModel<string[]>({ default: [] })
const error = defineModel<boolean>('error', { default: false })

const inputValue = ref<string>('')
const chipsWithErrors = ref<Map<string, number>>(new Map())

const hasErrors = computed(() => props.hasError || chipsWithErrors.value.size > 0)

watch(hasErrors, () => {
  emit('errorStateChanged', hasErrors.value)
  error.value = hasErrors.value
})

onMounted(() => {
  if (chips.value.length === 0)
    return

  chips.value.forEach((chip) => {
    if (!validateChip(chip)) {
      createErrorChip(chip)
    }
  })
})

function validateChip(chip: string): boolean {
  if (!props.validator)
    return true

  if (props.validator instanceof RegExp) {
    return props.validator.test(chip)
  }

  return props.validator(chip)
}

function handleKeyDown(e: KeyboardEvent) {
  const shouldCreateNewChip = props.chipSeparators.includes(e.key) && !!inputValue.value

  if (shouldCreateNewChip) {
    handleCreateChip(inputValue.value)
    inputValue.value = ''
  }
}

function handleCreateChip(chip: string) {
  const isValid = validateChip(chip)

  if (!isValid) {
    createErrorChip(chip)
  }

  chips.value = [...chips.value, chip]

  emit('chipsUpdated', chips.value)
}

function handleChipRemoval(chip: string, chipIndex: number) {
  chips.value = chips.value.toSpliced(chipIndex, 1)

  if (chipsWithErrors.value.has(chip)) {
    removeErrorChip(chip)
  }

  emit('chipsUpdated', chips.value)
}

function createErrorChip(chip: string) {
  const currentChipCount = chipsWithErrors.value.get(chip) ?? 0

  chipsWithErrors.value.set(chip, currentChipCount + 1)
}

function removeErrorChip(chip: string) {
  const currentChipCount = chipsWithErrors.value.get(chip)

  if (!currentChipCount)
    return

  if (currentChipCount === 1) {
    chipsWithErrors.value.delete(chip)

    return
  }

  chipsWithErrors.value.set(chip, currentChipCount - 1)
}
</script>
