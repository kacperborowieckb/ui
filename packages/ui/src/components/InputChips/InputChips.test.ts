import { shallowMount } from '@vue/test-utils'

import { Chip } from '../Chip'
import { InputBase } from '../InputBase'
import InputChips from './InputChips.vue'
import type { InputChipsProps } from './InputChips.vue'

describe('inputChips', () => {
  const getWrapper = (props?: InputChipsProps & { modelValue?: string[] }) => {
    return shallowMount(InputChips, {
      props,
      global: {
        stubs: {
          InputWrapper: {
            template: '<div data-testid="input-wrapper"><slot /></div>',
          },
        },
      },
    })
  }

  it('should render correctly', async () => {
    const wrapper = getWrapper({ chipSeparators: [','], modelValue: ['chip1', 'chip2'] })

    await wrapper.vm.$nextTick()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render correctly with error chips', async () => {
    const wrapper = getWrapper({
      chipSeparators: [','],
      modelValue: ['mo', 'mockChip2'],
      validator: val => val.length < 4,
    })

    await wrapper.vm.$nextTick()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should create a new chips correctly', async () => {
    const wrapper = getWrapper({
      chipSeparators: [',', 'Enter'],
    })

    const input = wrapper.findComponent(InputBase)

    await input.setValue('some')
    await input.trigger('keydown', { key: '.' })

    await input.trigger('keydown', { key: ',' })

    await input.setValue('some2')
    await input.trigger('keydown', { key: 'Enter' })

    expect(wrapper.emitted('update:modelValue'))
      .toStrictEqual([[['some']], [['some', 'some2']]])
  })

  it('should create error chips correctly', async () => {
    const wrapper = getWrapper({
      chipSeparators: [',', 'Enter'],
      validator: val => val !== 'some',
    })

    const input = wrapper.findComponent(InputBase)

    await input.setValue('some')
    await input.trigger('keydown', { key: ',' })

    await input.setValue('some')
    await input.trigger('keydown', { key: ',' })

    expect(wrapper.emitted('update:modelValue')).toStrictEqual([[['some']], [['some', 'some']]])
    expect(wrapper.emitted('update:error')).toStrictEqual([[true]])
    expect((wrapper.vm as any).chipsWithErrors).toStrictEqual(new Map([['some', 2]]))
  })

  it('should correctly handle regexp as validator', async () => {
    const wrapper = getWrapper({
      chipSeparators: [',', 'Enter'],
      validator: /^\d*$/,
    })

    const input = wrapper.findComponent(InputBase)

    await input.setValue('some')
    await input.trigger('keydown', { key: ',' })

    expect(wrapper.emitted('update:error')).toStrictEqual([[true]])
    expect((wrapper.vm as any).chipsWithErrors).toStrictEqual(new Map([['some', 1]]))
  })

  it('should correctly delete chip', () => {
    const wrapper = getWrapper({
      chipSeparators: [','],
      modelValue: ['mockChip', 'mockChip2'],
    })

    const lastChip = wrapper.findAllComponents(Chip).at(-1)

    lastChip?.vm.$emit('chipRemoval')

    expect(wrapper.emitted('update:modelValue')).toStrictEqual([[['mockChip']]])
  })

  it('should correctly delete error chip', async () => {
    const wrapper = getWrapper({
      chipSeparators: [','],
      modelValue: ['mockChip', 'mockChip2'],
      validator: val => val.includes('2'),
    })

    await wrapper.vm.$nextTick()

    const errorChip = wrapper.findAllComponents(Chip).find((chip) => {
      return chip.props().variant === 'error'
    })

    errorChip?.vm.$emit('chipRemoval')

    await wrapper.vm.$nextTick()

    expect(wrapper.emitted('update:error')).toStrictEqual([[true], [false]])
    expect(wrapper.emitted('update:modelValue')).toStrictEqual([[['mockChip2']]])
    expect((wrapper.vm as any).chipsWithErrors).toStrictEqual(new Map())
  })

  it('should correctly delete error chip while duplicate error', async () => {
    const wrapper = getWrapper({
      chipSeparators: [','],
      modelValue: ['mockChip', 'mockChip2', 'mockChip'],
      validator: val => val.includes('2'),
    })

    await wrapper.vm.$nextTick()

    const errorChip = wrapper.findAllComponents(Chip).find((chip) => {
      return chip.props().variant === 'error'
    })

    errorChip?.vm.$emit('chipRemoval')

    await wrapper.vm.$nextTick()

    expect(wrapper.emitted('update:error')).toStrictEqual([[true]])
    expect(wrapper.emitted('update:modelValue')).toStrictEqual([[['mockChip2', 'mockChip']]])
    expect((wrapper.vm as any).chipsWithErrors).toStrictEqual(new Map([['mockChip', 1]]))
  })
})
