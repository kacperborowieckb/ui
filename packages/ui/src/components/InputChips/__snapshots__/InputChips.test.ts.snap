// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`inputChips > should render correctly 1`] = `
"<div data-testid="input-wrapper" disabled="false" haserror="false">
  <chip-stub content="chip1" variant="default" focusable="true" disabled="false" withbutton="true"></chip-stub>
  <chip-stub content="chip2" variant="default" focusable="true" disabled="false" withbutton="true"></chip-stub>
  <input-base-stub modelvalue="" modelmodifiers="[object Object]" disabled="false"></input-base-stub>
</div>"
`;

exports[`inputChips > should render correctly with error chips 1`] = `
"<div data-testid="input-wrapper" disabled="false" haserror="true">
  <chip-stub content="mo" variant="default" focusable="true" disabled="false" withbutton="true"></chip-stub>
  <chip-stub content="mockChip2" variant="error" focusable="true" disabled="false" withbutton="true"></chip-stub>
  <input-base-stub modelvalue="" modelmodifiers="[object Object]" disabled="false"></input-base-stub>
</div>"
`;
